crates/outline_panel/src/outline_panel.rs

```diff
@@ -3267,6 +3373,57 @@ impl OutlinePanel {
                                         Some(UPDATE_DEBOUNCE)
                                     };
                                     excerpt.outlines = ExcerptOutlines::Outlines(fetched_outlines);
+
+                                    // Collect outlines to check for default expansion
+                                    let outlines_to_check =
+                                        if let Some(default_depth) = pending_default_depth {
+                                            if let ExcerptOutlines::Outlines(outlines) =
+                                                &excerpt.outlines
+                                            {
+                                                if default_depth == 0 {
+                                                    // For depth 0, collapse all outlines with children
+                                                    outlines.clone()
+                                                } else {
+                                                    // For other depths, collapse outlines at or below that depth
+                                                    outlines
+                                                        .iter()
+                                                        .filter(|outline| {
+                                                            outline.depth >= default_depth
+                                                        })
+                                                        .cloned()
+                                                        .collect::<Vec<_>>()
+                                                }
+                                            } else {
+                                                Vec::new()
+                                            }
+                                        } else {
+                                            Vec::new()
+                                        };
+
+                                    Some((debounce, outlines_to_check))
+                                } else {
+                                    None
+                                };
+
+                                // Now check which outlines have children and should be collapsed
+                                if let Some((debounce, outlines_to_check)) = outlines_to_check {
+                                    for outline in outlines_to_check {
+                                        let outline_entry = OutlineEntryOutline {
+                                            buffer_id,
+                                            excerpt_id,
+                                            outline: outline.clone(),
+                                        };
+                                        if outline_panel.has_outline_children(&outline_entry, cx) {
```

Comment:
This is odd: effectively, we go over all `fetched_outlines` and for each, we we-iteratethe outlines list again to find the children.

It's a plain O(N^2) for nothing and should be fixed.
We do `outlines.retain(|outline| {` right above and can prepare a depth map for any decision of a kind.